File: hash_maps/hash_maps.ts
/**
 * HASH MAPS  ·  TypeScript
 * Manual HashMap + frequency patterns + sliding window + prefix sum.
 */

class HashMap<K, V> {
  private buckets: [K, V][][];
  private cap: number;
  size = 0;
  constructor(cap = 16) { this.cap = cap; this.buckets = Array.from({length: cap}, () => []); }
  private h(key: K): number { return Math.abs(String(key).split('').reduce((a,c)=>((a<<5)-a)+c.charCodeAt(0),0)) % this.cap; }
  put(key: K, val: V): void {
    const b = this.buckets[this.h(key)];
    const i = b.findIndex(([k])=>k===key);
    if (i>=0) b[i][1]=val; else { b.push([key,val]); this.size++; }
  }
  get(key: K): V|undefined { return this.buckets[this.h(key)].find(([k])=>k===key)?.[1]; }
  remove(key: K): void {
    const b = this.buckets[this.h(key)]; const i = b.findIndex(([k])=>k===key);
    if (i>=0) { b.splice(i,1); this.size--; }
  }
  has(key: K): boolean { return this.get(key)!==undefined; }
}

// ── Problems ──────────────────────────────────
function groupAnagrams(strs: string[]): string[][] {
  const m = new Map<string,string[]>();
  for (const s of strs) { const k=[...s].sort().join(''); if(!m.has(k))m.set(k,[]); m.get(k)!.push(s); }
  return [...m.values()];
}

function longestConsecutive(nums: number[]): number {
  const s = new Set(nums); let best = 0;
  for (const n of s) { if (!s.has(n-1)) { let cur=n,len=1; while(s.has(cur+1)){cur++;len++;} best=Math.max(best,len); } }
  return best;
}

function lengthOfLongestSubstring(s: string): number {
  const last = new Map<string,number>(); let best=0,left=0;
  for (let right=0;right<s.length;right++) {
    const c=s[right];
    if(last.has(c)&&last.get(c)!>=left) left=last.get(c)!+1;
    last.set(c,right); best=Math.max(best,right-left+1);
  }
  return best;
}

function minWindowSubstring(s: string, t: string): string {
  const need = new Map<string,number>(); for(const c of t) need.set(c,(need.get(c)??0)+1);
  let missing=t.length, left=0, best='';
  for (let right=0;right<s.length;right++) {
    const c=s[right]; if((need.get(c)??0)>0) missing--;
    need.set(c,(need.get(c)??0)-1);
    if(missing===0){
      while((need.get(s[left])??0)<0){need.set(s[left],(need.get(s[left])??0)+1);left++;}
      if(!best||right-left+1<best.length) best=s.slice(left,right+1);
      need.set(s[left],(need.get(s[left])??0)+1); missing++; left++;
    }
  }
  return best;
}

function subarraySumEqualsK(nums: number[], k: number): number {
  let count=0,prefix=0; const freq=new Map([[0,1]]);
  for(const n of nums){prefix+=n;count+=(freq.get(prefix-k)??0);freq.set(prefix,(freq.get(prefix)??0)+1);}
  return count;
}

function wordPattern(pattern: string, s: string): boolean {
  const words=s.split(' '); if(pattern.length!==words.length) return false;
  const pw=new Map<string,string>(), wp=new Map<string,string>();
  for(let i=0;i<pattern.length;i++){
    const [p,w]=[pattern[i],words[i]];
    if(pw.has(p)&&pw.get(p)!==w) return false;
    if(wp.has(w)&&wp.get(w)!==p) return false;
    pw.set(p,w); wp.set(w,p);
  }
  return true;
}

function isIsomorphic(s: string, t: string): boolean {
  const st=new Map<string,string>(), ts=new Map<string,string>();
  for(let i=0;i<s.length;i++){
    if((st.get(s[i])??t[i])!==t[i]||(ts.get(t[i])??s[i])!==s[i]) return false;
    st.set(s[i],t[i]); ts.set(t[i],s[i]);
  }
  return true;
}

function firstUniqChar(s: string): number {
  const cnt=new Map<string,number>(); for(const c of s) cnt.set(c,(cnt.get(c)??0)+1);
  for(let i=0;i<s.length;i++) if(cnt.get(s[i])===1) return i;
  return -1;
}

function findAllAnagrams(s: string, p: string): number[] {
  const need=new Map<string,number>(); for(const c of p) need.set(c,(need.get(c)??0)+1);
  const window=new Map<string,number>(); const result: number[]=[]; const k=p.length;
  const mEq=(a:Map<string,number>,b:Map<string,number>)=>a.size===b.size&&[...a].every(([k,v])=>b.get(k)===v);
  for(let i=0;i<s.length;i++){
    const c=s[i]; window.set(c,(window.get(c)??0)+1);
    if(i>=k){const lc=s[i-k];const nc=(window.get(lc)??0)-1;nc===0?window.delete(lc):window.set(lc,nc);}
    if(mEq(window,need)) result.push(i-k+1);
  }
  return result;
}

// ── Tests ─────────────────────────────────────
function assert(c: boolean, m: string){if(!c)throw new Error(`FAIL: ${m}`);}
const eq=(a:unknown,b:unknown)=>JSON.stringify(a)===JSON.stringify(b);

function runTests(){
  console.log("Running hash map tests...\n");

  const hm=new HashMap<string,number>();
  hm.put('a',1); hm.put('b',2); hm.put('a',99);
  assert(hm.get('a')===99&&hm.get('b')===2,"put/get/update");
  assert(hm.get('z')===undefined,"get missing");
  hm.remove('a'); assert(!hm.has('a'),"remove");
  console.log("  ✅ HashMap: put / get / update / remove");

  const groups=groupAnagrams(["eat","tea","tan","ate","nat","bat"]);
  assert(groups.length===3,"groupAnagrams count");
  console.log("  ✅ groupAnagrams");

  assert(longestConsecutive([100,4,200,1,3,2])===4,"longCons");
  assert(longestConsecutive([0,3,7,2,5,8,4,6,0,1])===9,"longCons2");
  console.log("  ✅ longestConsecutive");

  assert(lengthOfLongestSubstring("abcabcbb")===3,"longestSub");
  assert(lengthOfLongestSubstring("bbbbb")===1,"longestSub2");
  console.log("  ✅ lengthOfLongestSubstring");

  assert(minWindowSubstring("ADOBECODEBANC","ABC")==="BANC","minWindow");
  assert(minWindowSubstring("a","a")==="a","minWindow2");
  assert(minWindowSubstring("a","b")==="","minWindow3");
  console.log("  ✅ minWindowSubstring");

  assert(subarraySumEqualsK([1,1,1],2)===2,"subarraySum");
  assert(subarraySumEqualsK([1,2,3],3)===2,"subarraySum2");
  console.log("  ✅ subarraySumEqualsK");

  assert(wordPattern("abba","dog cat cat dog"),"wordPat");
  assert(!wordPattern("abba","dog cat cat fish"),"wordPat2");
  console.log("  ✅ wordPattern");

  assert(isIsomorphic("egg","add")&&!isIsomorphic("foo","bar"),"isomorph");
  console.log("  ✅ isIsomorphic");

  assert(firstUniqChar("leetcode")===0,"firstUniq");
  assert(firstUniqChar("aabb")===-1,"firstUniq2");
  console.log("  ✅ firstUniqChar");

  assert(eq(findAllAnagrams("cbaebabacd","abc"),[0,6]),"anagrams");
  assert(eq(findAllAnagrams("abab","ab"),[0,1,2]),"anagrams2");
  console.log("  ✅ findAllAnagrams");

  console.log("\n🎉 All hash map tests passed!");
}
runTests();
export { HashMap, groupAnagrams, longestConsecutive, lengthOfLongestSubstring, minWindowSubstring, subarraySumEqualsK, wordPattern, isIsomorphic, firstUniqChar, findAllAnagrams };

File: 02_hash_maps/hash_maps.rs
//! HASH MAPS  ·  Rust
//! HashMap<K,V> from std::collections. O(1) avg get/insert/remove.
//! HashMap is unordered. For ordered, use BTreeMap.
use std::collections::{HashMap, HashSet};

// ── Problems ──────────────────────────────────
pub fn group_anagrams(strs: &[&str]) -> Vec<Vec<String>> {
    let mut map: HashMap<Vec<char>, Vec<String>> = HashMap::new();
    for &s in strs {
        let mut key: Vec<char> = s.chars().collect(); key.sort();
        map.entry(key).or_default().push(s.to_string());
    }
    map.into_values().collect()
}

pub fn longest_consecutive(nums: &[i32]) -> usize {
    let s: HashSet<i32> = nums.iter().copied().collect();
    let mut best = 0;
    for &n in &s {
        if !s.contains(&(n-1)) {
            let mut cur = n; let mut len = 1;
            while s.contains(&(cur+1)) { cur+=1; len+=1; }
            best = best.max(len);
        }
    }
    best
}

pub fn length_of_longest_substring(s: &str) -> usize {
    let chars: Vec<char> = s.chars().collect();
    let mut last: HashMap<char,usize> = HashMap::new();
    let mut best = 0; let mut left = 0;
    for (right, &c) in chars.iter().enumerate() {
        if let Some(&prev) = last.get(&c) { if prev >= left { left = prev + 1; } }
        last.insert(c, right);
        best = best.max(right - left + 1);
    }
    best
}

pub fn min_window_substring<'a>(s: &'a str, t: &str) -> &'a str {
    let sc: Vec<char> = s.chars().collect();
    let mut need: HashMap<char,i32> = HashMap::new();
    for c in t.chars() { *need.entry(c).or_insert(0) += 1; }
    let mut missing = t.len() as i32;
    let (mut left, mut best_left, mut best_len) = (0usize, 0, usize::MAX);
    for (right, &c) in sc.iter().enumerate() {
        if *need.get(&c).unwrap_or(&0) > 0 { missing -= 1; }
        *need.entry(c).or_insert(0) -= 1;
        if missing == 0 {
            while *need.get(&sc[left]).unwrap_or(&0) < 0 { *need.entry(sc[left]).or_insert(0) += 1; left += 1; }
            if right - left + 1 < best_len { best_len = right-left+1; best_left = left; }
            *need.entry(sc[left]).or_insert(0) += 1; missing += 1; left += 1;
        }
    }
    if best_len == usize::MAX { "" } else { &s[best_left..best_left+best_len] }
}

pub fn subarray_sum_k(nums: &[i32], k: i32) -> i32 {
    let mut count = 0; let mut prefix = 0;
    let mut freq: HashMap<i32,i32> = HashMap::from([(0,1)]);
    for &n in nums {
        prefix += n;
        count += freq.get(&(prefix-k)).copied().unwrap_or(0);
        *freq.entry(prefix).or_insert(0) += 1;
    }
    count
}

pub fn word_pattern(pattern: &str, s: &str) -> bool {
    let words: Vec<&str> = s.split_whitespace().collect();
    if pattern.len() != words.len() { return false; }
    let mut pw: HashMap<char,&str> = HashMap::new();
    let mut wp: HashMap<&str,char> = HashMap::new();
    for (p, w) in pattern.chars().zip(words.iter()) {
        if pw.get(&p).map_or(false, |&v| v != *w) { return false; }
        if wp.get(w).map_or(false, |&v| v != p) { return false; }
        pw.insert(p, w); wp.insert(w, p);
    }
    true
}

pub fn first_unique_char(s: &str) -> i32 {
    let mut cnt: HashMap<char,usize> = HashMap::new();
    for c in s.chars() { *cnt.entry(c).or_insert(0) += 1; }
    for (i, c) in s.chars().enumerate() { if cnt[&c] == 1 { return i as i32; } }
    -1
}

pub fn find_all_anagrams(s: &str, p: &str) -> Vec<usize> {
    let sc: Vec<char> = s.chars().collect();
    let k = p.len();
    let mut need: HashMap<char,i32> = HashMap::new();
    for c in p.chars() { *need.entry(c).or_insert(0) += 1; }
    let mut window: HashMap<char,i32> = HashMap::new();
    let mut result = vec![];
    for (i, &c) in sc.iter().enumerate() {
        *window.entry(c).or_insert(0) += 1;
        if i >= k { let lc=sc[i-k]; let v=window.entry(lc).or_insert(0); *v-=1; if *v==0 { window.remove(&lc); } }
        if i + 1 >= k && window == need { result.push(i + 1 - k); }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_group_anagrams() {
        let r = group_anagrams(&["eat","tea","tan","ate","nat","bat"]);
        assert_eq!(r.len(), 3);
    }
    #[test]
    fn test_longest_consecutive() {
        assert_eq!(longest_consecutive(&[100,4,200,1,3,2]), 4);
        assert_eq!(longest_consecutive(&[0,3,7,2,5,8,4,6,0,1]), 9);
    }
    #[test]
    fn test_longest_substring() {
        assert_eq!(length_of_longest_substring("abcabcbb"), 3);
        assert_eq!(length_of_longest_substring("bbbbb"), 1);
    }
    #[test]
    fn test_min_window() {
        assert_eq!(min_window_substring("ADOBECODEBANC","ABC"), "BANC");
        assert_eq!(min_window_substring("a","a"), "a");
        assert_eq!(min_window_substring("a","b"), "");
    }
    #[test]
    fn test_subarray_sum_k() {
        assert_eq!(subarray_sum_k(&[1,1,1],2), 2);
        assert_eq!(subarray_sum_k(&[1,2,3],3), 2);
    }
    #[test]
    fn test_word_pattern() {
        assert!(word_pattern("abba","dog cat cat dog"));
        assert!(!word_pattern("abba","dog cat cat fish"));
    }
    #[test]
    fn test_first_unique() {
        assert_eq!(first_unique_char("leetcode"), 0);
        assert_eq!(first_unique_char("aabb"), -1);
    }
    #[test]
    fn test_find_all_anagrams() {
        assert_eq!(find_all_anagrams("cbaebabacd","abc"), vec![0,6]);
        assert_eq!(find_all_anagrams("abab","ab"), vec![0,1,2]);
    }
}
